### PR TITLE
Unet integration

### PR DESCRIPTION
**Pull request by KorenMary**

### CODE REVIEW

File: src/client/dcp_client/config.cfg
Comment (christinab12) [inline src/client/dcp_client/config.cfg:1]: can you remove this? you can add your configs to your gitignore so we you don't accidentaly commit changes to these files when playing around withe the configurations

File: src/client/dcp_client/gui/napari_window.py
Comment (christinab12) [inline src/client/dcp_client/gui/napari_window.py:1]: are these changes part of the work for the front-end extension to work with multiple masks?

File: src/server/dcp_server/config.cfg
Comment (christinab12) [inline src/server/dcp_server/config.cfg:1]: rename unet to end-2-end

File: src/server/dcp_server/models.py [hunk 1]
```diff
@@ -45,8 +48,10 @@ def eval(self, img):
         :return: mask of the image, list of 2D arrays, or single 3D array (if do_3D=True) labelled image.
         :rtype: np.ndarray
         """  
-        return super().eval(x=img, **self.eval_config["segmentor"])[0] # 0 to take only mask
```
Comment (christinab12) [inline src/server/dcp_server/models.py:48]: are changes to ```CustomCellposeModel``` necessary for Unet integration?

File: src/server/dcp_server/test_config.cfg
Comment (christinab12) [inline src/server/dcp_server/test_config.cfg:1]: move to tests dir

File: src/server/test/test_integration.py
Comment (christinab12) [inline src/server/test/test_integration.py:1]: * these changes should be part of ```reload saved model``` branch?
* is unet now included in the tests?

File: src/server/dcp_server/models.py [hunk 2]
```diff
@@ -295,3 +297,164 @@ def eval(self, img):
...
+        train_dataloader = DataLoader(train_dataset, batch_size=batch_size)
+
+        loss_fn = nn.CrossEntropyLoss()
+        optimizer = Adam(params=self.parameters(), lr=lr)
+
+        for _ in tqdm(range(epochs), desc="Running UNet training"):
+
+            self.loss = 0
+
+            for _, (imgs, masks) in enumerate(train_dataloader):
```
Comment (christinab12) [inline src/server/dcp_server/models.py:423]: why do enumerate if you are not going to use the index? You can change to: ```for imgs, masks in train_dataloader:```

File: src/server/dcp_server/models.py [hunk 3]
```diff
@@ -295,3 +297,164 @@ def eval(self, img):
...
+                #forward path
+                preds = self.forward(imgs)
+                loss = loss_fn(preds, masks)
+                
+                #backward path
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+            
+            self.loss /= len(train_dataloader)
```
Comment (christinab12) [inline src/server/dcp_server/models.py:436]: you are not adding something to self.loss, only computing loss in the loop

File: src/server/dcp_server/models.py [hunk 4]
```diff
@@ -295,3 +297,164 @@ def eval(self, img):
...
+            Output: y_hat - The predicted label
+        """ 
+        # normalise
+        img = (img-np.min(img))/(np.max(img)-np.min(img))
+        img = torch.from_numpy(img).float().unsqueeze(0)
+
+        img = img.unsqueeze(1) if img.ndim == 3 else img
+        # convert to tensor
+        # img = torch.permute(torch.tensor(img.astype(np.float32)), (2, 0, 1)).unsqueeze(0) 
+
```
Comment (christinab12) [inline src/server/dcp_server/models.py:452]: add model.eval or torch.no_grad?

File: src/server/dcp_server/models.py [hunk 5]
```diff
@@ -295,3 +297,164 @@ def eval(self, img):
...
+        preds = self.forward(img)
+        y_labels = torch.argmax(preds, 1).numpy()[0]
+
+        y_instances = label((y_labels > 0).astype(int))[0]
+
+        y_hat = np.stack((y_instances, y_labels))
+
+        return y_hat
```
Comment (christinab12) [inline src/server/dcp_server/models.py:460]: can you rename these variables (y_hat, y_instances etc.) to the variable names we used in CellPosePatchCNN, so we are consistent in our naming?

**Review by christinab12 — CHANGES_REQUESTED** [review]
Hi @KorenMary, please take a look at my comments, some small changes are needed but all in all it is looking good thanks :)

File: src/server/test/test_integration.py
Comment (KorenMary) [reply src/server/test/test_integration.py:1]: Yes, tests for UNet are also included in this branch should I add them in the integration-tests branch?

File: src/server/dcp_server/models.py (hunk 2 above)
Comment (KorenMary) [reply src/server/dcp_server/models.py:423]: I agree, I was using index to print, but now it's not necessary any more, thanks.

File: src/server/test/test_integration.py
Comment (christinab12) [reply src/server/test/test_integration.py:1]: no here is good thanks. Every time we add a new model, we should also make sure test for this are also running, that's all!
